### ISSUE TITLE
[DISCUSSION] How to deal with limited annotation support in `dask-expr`

### ISSUE THREAD

**Issue by hendrikmakait** [dask/dask#10937]
At the moment, `dask-expr` has only limited support for annotations (https://github.com/dask-contrib/dask-expr/issues/13). Specifically, it only supports annotations at graph submission. For example, you can restrict resources for an *entire* computation with

```python
# Define your entire computation graph here
df = ...

with dask.annotate(resources={'GPU': 1}):
    # Submit the graph with resource restrictions  
    res = client.compute(df)

```

Applying resource restrictions at graph creation does not work and will be silently ignored:

```python
with dask.annotate(resources={'GPU': 1}):
    # Define a computation that should run on GPUs here
    df = ...

# Submitting the graph outside of the annotation context will ignore the annotations.
client.compute(df)
```

Before choosing `dask-expr` as the default mechanism (#10934), we should, therefore, reach a consensus on how we want to deal with this lack of support.

Primarily, we need to answer the question: Are we okay with local annotations breaking silently when defaulting to `dask-expr`? If not, what are the minimal requirements that we need to support before the switch?

Generally, there is some contention on what annotation support would ideally look like. The current implementation falls short in multiple regards, e.g., by losing annotations in various codepaths, or mixing no clear separation between loose) 8 worker restrictions  Some points that have been raised here and can start a discussion are:

- Do we want general-purpose mechanism for all kinds of annotations or should we instead limit this functionality to multiple explicit APIs, e.g., for restricting tasks to workers or resources, and attaching metadata to tasks?
- How do annotations interact with logical plan optimization, e.g., plan rewrites, plan reordering, and fusion?
- What annotation functionality is even used in practice?

**Comment by mrocklin** [comment]
My guess is that query optimization will make more users happy than losing annotations will make users sad.  Because of this, I'm in favor of merging in query optimization even if it disrupts annotations.  I think that it's ok to make some existing users sad.

Of course I'd be happy to see a better solution here, but I wouldn't want this to slow things down much.

**Comment by hendrikmakait** [comment]
I agree that query optimization will be of much bigger benefit than annotations for most users. However, I don't think breaking things silently will be a good UX for those users who are affected. To move forward quickly, I think that adding a warning to `dask.annotate`, which contains more information, might be a nice compromise. Similar to #10925, users should be able to silence this warning.

**Comment by mrocklin** [comment]
Yup, I don't mind seeing improvements to annotations UX.  However, I don't think that the release of query planning should be slowed down by this.  I agree with scope, I wouldn't agree with a blocking change (not that you're proposing that) and I'm not sure how highly I would prioritize this (I don't see many users with annotations in dataframes).  However, if you have different knowledge (for example if there has been evidence that people use annotations with dask.dataframe) then I could agree that it's worth prioritizing.  If there has not been evidence that this is in use then I probably wouldn't bother (we have plenty of other high priority work to do) unless it's very easy.

**Comment by rjzamora** [comment]
I'm interested in picking up this feature (or something similar) if someone isn't already working on it. I'm seeing a lot of internal interest in using different resources to execute different expressions/tasks within the same query. The clearest motivation is to use a large number of CPU-only workers to drive remote-IO tasks, while a smaller number of GPU workers are used for compute-heavy tasks (including training and batched inference).

My understanding is that there are two parts to the annotations problem:

1. We need a way to attach annotations to specific expressions
2. We need a way to communicate the per-expression (or per-task) annotations to the scheduler

I think it makes sense to answer (2) first.

The “legacy” design relies on the fact that HighLevelGraph objects can be communicated to the scheduler, and distinct annotations can be stored on each Layer in the HighLevelGraph. At the moment , every graph is converted to a HighLevelGraph before it is communicated to the scheduler. However, it seems pretty clear that we eventually want to get rid of the HighLevelGraph/Layer infrastructure altogether.

Since we are moving away from the HighLevelGraph/Layer design either way. I suppose there are two general designs that can be used to communicate annotations that target a finer grain than the entire graph:

- (A) We pickle/send the `Expr` object to the scheduler in the same way that we are currently doing for HighLevelGraph objects

- (B) We continue materializing `Expr`-based graphs on the client side, and introduce a new mechanism to send separate annotation information

@fjetter - I believe you spent some time exploring (A) - Is that right?

**Comment by fjetter** [comment]
> My understanding is that there are two parts to the annotations problem:
> 
> We need a way to attach annotations to specific expressions
> We need a way to communicate the per-expression (or per-task) annotations to the scheduler

This is not what stopped implementation but rather that it is unclear how annotations are supposed to propagate through the optimizer.
The serialization of expressions is something I already did on some feature branches but we never merged this due to some perf problems and because the benefits were marginal for most cases.

The optimizer is the much tougher problem. After plenty of conversations about this and similar topics, I'm not strongly motivated to reintroduce "generic, whatever you think" annotations as they are currently supported. 
For example, you are looking for Restrictions or Resources. We might want to implement this as a special case. This will likely be handled in a different way than Retries or priorities (If we even want to continue to support priotities as annotations)

**Comment by rjzamora** [comment]
>The optimizer is the much tougher problem. After plenty of conversations about this and similar topics, I'm not strongly motivated to reintroduce "generic, whatever you think" annotations as they are currently supported.
For example, you are looking for Restrictions or Resources. We might want to implement this as a special case. This will likely be handled in a different way than Retries or priorities (If we even want to continue to support priotities as annotations)

I'm pretty sure I agree with you. I honestly don't care about supporting the "generic, whatever you think" annotations in an arbitrary context block. Rather, I want to make it possible for: `Expr._layer` to define resource restrictions when the tasks are generated for that particular expression (or *something* like this). I'd be happy to break this off from general `annotations` support. If we do this, we still need to decide how we want to communicate this information to the scheduler.

**Comment by rjzamora** [comment]
@fjetter @phofl - I thought through this problem a bit and decided to propose https://github.com/dask/dask-expr/pull/1116

I'll be happy to revise the idea (and completely throw it out if necessary), but I'm hopeful we can find an API that is both maintainable and functional.
